File: catapult/javascript/06_aggregate_complete_transaction_mosaic.js
const {
  Account,
  TransferTransaction,
  AggregateTransaction,
  Deadline,
  PlainMessage,
  RepositoryFactoryHttp,
  TransactionService,
  NamespaceId,
  Mosaic,
  UInt64,
} = require("@tech-bureau/symbol-sdk");
const config = require("config");
const { firstValueFrom } = require("rxjs");

const fetch = require("node-fetch");
const fetchCookie = require("fetch-cookie");
const toughCookie = require("tough-cookie");

const url = config.get("private1.url");
const rawprivatekey_w = config.get("private1.workaddress.privatekey");
const rawprivatekey_1 = config.get("private1.address1.privatekey");
const rawprivatekey_2 = config.get("private1.address2.privatekey");
const rawprivatekey_3 = config.get("private1.address3.privatekey");
const rawprivatekey_4 = config.get("private1.address4.privatekey");
const rawprivatekey_5 = config.get("private1.address5.privatekey");
const rawprivatekey_6 = config.get("private1.address6.privatekey");
const minfeemultiplier = config.get("private1.minfeemultiplier");

//送信するMosaicのNamespace
const namespace = config.get("private1.workaddress.namespace");

// 送信するMosaic数
const sendmosaicamount = 1;

(async () => {
  // 一回mijinにアクセスして、Cookieを取得する
  const cookieJar = new toughCookie.CookieJar();
  const fetchCookieJar = fetchCookie(fetch, cookieJar);
  await fetchCookieJar(url);
  const cookie = await cookieJar.getCookieString(url);

  // Websocketオプション Cookieを設定する
  const websocketOptions = {
    headers: { cookie: cookie },
  };

  const repo = new RepositoryFactoryHttp(url, {
    fetchApi: fetchCookieJar,
    websocketOptions: websocketOptions,
  });

  //Networktypeをcatapultから取得
  const networkType = await firstValueFrom(repo.getNetworkType());
  //GenerationHashをcatapultから取得
  const ghash = await firstValueFrom(repo.getGenerationHash());

  console.log("network", networkType);
  console.log("GenerationHash", ghash);

  const epoch = await firstValueFrom(repo.getEpochAdjustment());

  //送信元のアカウント(workaddress)をセット
  const account_w = Account.createFromPrivateKey(rawprivatekey_w, networkType);

  //アカウント1-6をセット
  const account_1 = Account.createFromPrivateKey(rawprivatekey_1, networkType);
  const account_2 = Account.createFromPrivateKey(rawprivatekey_2, networkType);
  const account_3 = Account.createFromPrivateKey(rawprivatekey_3, networkType);
  const account_4 = Account.createFromPrivateKey(rawprivatekey_4, networkType);
  const account_5 = Account.createFromPrivateKey(rawprivatekey_5, networkType);
  const account_6 = Account.createFromPrivateKey(rawprivatekey_6, networkType);

  const namespaceId = new NamespaceId(namespace);
  const mosaic = new Mosaic(namespaceId, UInt64.fromUint(sendmosaicamount));

  //トランザクションを作成するための内容を作る
  const transferTransaction1 = TransferTransaction.create(
    Deadline.create(epoch), //いつまでトランザクションをもたせるか(default ２時間)
    account_1.address, //送信先アドレス1
    [mosaic], //送信するモザイク
    PlainMessage.create("TEST MIJIN tx1"), //メッセージ
    networkType
  );

  //トランザクションを作成するための内容を作る
  const transferTransaction2 = TransferTransaction.create(
    Deadline.create(epoch), //いつまでトランザクションをもたせるか(default ２時間)
    account_2.address, //送信先アドレス
    [mosaic], //送信するモザイク
    PlainMessage.create("TEST MIJIN tx2"), //メッセージ
    networkType
  );

  //トランザクションを作成するための内容を作る
  const transferTransaction3 = TransferTransaction.create(
    Deadline.create(epoch), //いつまでトランザクションをもたせるか(default ２時間)
    account_3.address, //送信先アドレス
    [mosaic], //送信するモザイク
    PlainMessage.create("TEST MIJIN tx3"), //メッセージ
    networkType
  );

  //トランザクションを作成するための内容を作る
  const transferTransaction4 = TransferTransaction.create(
    Deadline.create(epoch), //いつまでトランザクションをもたせるか(default ２時間)
    account_4.address, //送信先アドレス
    [mosaic], //送信するモザイク
    PlainMessage.create("TEST MIJIN tx4"), //メッセージ
    networkType
  );

  //トランザクションを作成するための内容を作る
  const transferTransaction5 = TransferTransaction.create(
    Deadline.create(epoch), //いつまでトランザクションをもたせるか(default ２時間)
    account_5.address, //送信先アドレス
    [mosaic], //送信するモザイク
    PlainMessage.create("TEST MIJIN tx5"), //メッセージ
    networkType
  );

  //トランザクションを作成するための内容を作る
  const transferTransaction6 = TransferTransaction.create(
    Deadline.create(epoch), //いつまでトランザクションをもたせるか(default ２時間)
    account_6.address, //送信先アドレス
    [mosaic], //送信するモザイク
    PlainMessage.create("TEST MIJIN tx6"), //メッセージ
    networkType
  );

  // トランザクションを纏め、AggregateTransactionを作成する。
  // (送信元が同じアドレスの処理なのでCompleteを使用)
  const aggregteTransaction = AggregateTransaction.createComplete(
    Deadline.create(epoch), //いつまでトランザクションをもたせるか(default ２時間),
    [
      transferTransaction1.toAggregate(account_w.publicAccount),
      transferTransaction2.toAggregate(account_w.publicAccount),
      transferTransaction3.toAggregate(account_w.publicAccount),
      transferTransaction4.toAggregate(account_w.publicAccount),
      transferTransaction5.toAggregate(account_w.publicAccount),
      transferTransaction6.toAggregate(account_w.publicAccount),
    ],
    networkType,
    []
  ).setMaxFeeForAggregate(minfeemultiplier, 0);

  //作成したトランザクションに署名する
  const signedTransaction = account_w.sign(aggregteTransaction, ghash);

  //この中にトランザクションIDが入ってる
  console.log("------------------- signedTransaction ------------------------");
  console.log(signedTransaction);
  console.log("------------------- sendTransaction ------------------------");

  //websocketを作成する
  const listener = repo.createListener();
  const transactionService = new TransactionService(repo.createTransactionRepository(), repo.createReceiptRepository());

  try {
    // websocketをopenする
    await listener.open();
    const tx = await firstValueFrom(transactionService.announce(signedTransaction, listener));
    console.log("------------------- Success --------------------------------");
    console.log("Success: Transaction", tx);
    console.log("------------------------------------------------------------");
    listener.close();
    return tx;
  } catch (e) {
    listener.close();
    e.code ? console.log("Failed TransactionStatusError", e.code) : console.log("Error: ", e);
    throw new Error(e);
  }
})();

File: catapult/python/06_aggregate_complete_transaction_mosaic.py
import datetime
from binascii import unhexlify
from json import dumps, loads

import httpx
import websocket
from helper import encode_message, generate_deadline, get_config
from symbolchain.CryptoTypes import Hash256, PrivateKey
from symbolchain.facade.SymbolFacade import SymbolFacade
from symbolchain.sc import Amount
from symbolchain.symbol.Network import Network


def aggregate_complete_transaction_mosaic():
    config = get_config()
    url = config["private1"]["url"]
    raw_privatekey_w = config["private1"]["workaddress"]["privatekey"]
    raw_privatekey_1 = config["private1"]["address1"]["privatekey"]
    raw_privatekey_2 = config["private1"]["address2"]["privatekey"]
    raw_privatekey_3 = config["private1"]["address3"]["privatekey"]
    raw_privatekey_4 = config["private1"]["address4"]["privatekey"]
    raw_privatekey_5 = config["private1"]["address5"]["privatekey"]
    raw_privatekey_6 = config["private1"]["address6"]["privatekey"]
    min_fee_multiplier = config["private1"]["minfeemultiplier"]
    mosaic_id = config["private1"]["workaddress"]["mosaicId"]

    # mijinにアクセスして、Cookie, GenerationHash, epochAdjustmentを取得する
    r = httpx.get(url + "/network/properties")
    cookies = r.cookies
    cookies_list = []
    for k, v in cookies.items():
        cookie_str = f"{k}={v}"
        cookies_list.append(cookie_str)
    cookie = "; ".join(f"{c}" for c in cookies_list)

    network = r.json()
    generation_hash = network["network"]["generationHashSeed"]
    epoch = network["network"]["epochAdjustment"]
    epoch = epoch.replace("s", "")

    facade = SymbolFacade("mijinnet")
    facade.network = Network(
        "mijinnet",
        0x60,
        datetime.datetime.fromtimestamp(int(epoch)),
        Hash256(generation_hash),
    )

    # 送信元のアカウントをセット
    account_w = facade.KeyPair(PrivateKey(unhexlify(raw_privatekey_w)))
    account_1 = facade.KeyPair(PrivateKey(unhexlify(raw_privatekey_1)))
    account_2 = facade.KeyPair(PrivateKey(unhexlify(raw_privatekey_2)))
    account_3 = facade.KeyPair(PrivateKey(unhexlify(raw_privatekey_3)))
    account_4 = facade.KeyPair(PrivateKey(unhexlify(raw_privatekey_4)))
    account_5 = facade.KeyPair(PrivateKey(unhexlify(raw_privatekey_5)))
    account_6 = facade.KeyPair(PrivateKey(unhexlify(raw_privatekey_6)))

    address_w = facade.network.public_key_to_address(account_w.public_key)
    address_1 = facade.network.public_key_to_address(account_1.public_key)
    address_2 = facade.network.public_key_to_address(account_2.public_key)
    address_3 = facade.network.public_key_to_address(account_3.public_key)
    address_4 = facade.network.public_key_to_address(account_4.public_key)
    address_5 = facade.network.public_key_to_address(account_5.public_key)
    address_6 = facade.network.public_key_to_address(account_6.public_key)

    # トランザクションを作成するための内容を作る
    deadline = generate_deadline(epoch)
    fee = Amount(min_fee_multiplier)
    amount = Amount(1)

    transfer_transaction1 = facade.transaction_factory.create_embedded(
        {
            "signer_public_key": account_w.public_key,
            "type": "transfer_transaction_v1",
            "recipient_address": address_1,
            "mosaics": [
                {
                    "mosaic_id": int(mosaic_id, 16),
                    "amount": amount,
                }
            ],
            "message": encode_message("TEST MIJIN tx1"),
        }
    )

    transfer_transaction2 = facade.transaction_factory.create_embedded(
        {
            "signer_public_key": account_w.public_key,
            "type": "transfer_transaction_v1",
            "recipient_address": address_2,
            "mosaics": [
                {
                    "mosaic_id": int(mosaic_id, 16),
                    "amount": amount,
                }
            ],
            "message": encode_message("TEST MIJIN tx2"),
        }
    )

    transfer_transaction3 = facade.transaction_factory.create_embedded(
        {
            "signer_public_key": account_w.public_key,
            "type": "transfer_transaction_v1",
            "recipient_address": address_3,
            "mosaics": [
                {
                    "mosaic_id": int(mosaic_id, 16),
                    "amount": amount,
                }
            ],
            "message": encode_message("TEST MIJIN tx3"),
        }
    )

    transfer_transaction4 = facade.transaction_factory.create_embedded(
        {
            "signer_public_key": account_w.public_key,
            "type": "transfer_transaction_v1",
            "recipient_address": address_4,
            "mosaics": [
                {
                    "mosaic_id": int(mosaic_id, 16),
                    "amount": amount,
                }
            ],
            "message": encode_message("TEST MIJIN tx4"),
        }
    )

    transfer_transaction5 = facade.transaction_factory.create_embedded(
        {
            "signer_public_key": account_w.public_key,
            "type": "transfer_transaction_v1",
            "recipient_address": address_5,
            "mosaics": [
                {
                    "mosaic_id": int(mosaic_id, 16),
                    "amount": amount,
                }
            ],
            "message": encode_message("TEST MIJIN tx5"),
        }
    )

    transfer_transaction6 = facade.transaction_factory.create_embedded(
        {
            "signer_public_key": account_w.public_key,
            "type": "transfer_transaction_v1",
            "recipient_address": address_6,
            "mosaics": [
                {
                    "mosaic_id": int(mosaic_id, 16),
                    "amount": amount,
                }
            ],
            "message": encode_message("TEST MIJIN tx6"),
        }
    )

    # トランザクションを纏める
    embedded_transactions = [
        transfer_transaction1,
        transfer_transaction2,
        transfer_transaction3,
        transfer_transaction4,
        transfer_transaction5,
        transfer_transaction6,
    ]
    transactions_hash = facade.hash_embedded_transactions(embedded_transactions)
    aggregate_transaction = facade.transaction_factory.create(
        {
            "type": "aggregate_complete_transaction_v2",
            "signer_public_key": account_w.public_key,
            "deadline": deadline,
            "transactions_hash": transactions_hash,
            "transactions": embedded_transactions,
            "fee": fee,
        }
    )
    aggregate_transaction.fee = Amount(0)

    # 作成したトランザクションに署名する
    signature = facade.sign_transaction(account_w, aggregate_transaction)
    json_payload = facade.transaction_factory.attach_signature(
        aggregate_transaction, signature
    )
    data = loads(json_payload)
    transaction_hash = facade.hash_transaction(aggregate_transaction)
    signed_transaction = {
        "SignedTransaction": {
            "payload": data["payload"],
            "hash": str(transaction_hash),
            "signerPublicKey": str(account_w.public_key),
            "type": aggregate_transaction.type_.value,
            "networkType": facade.network.identifier,
        }
    }

    # トランザクションID
    print("------------------- signedTransaction ------------------------")
    print(dumps(signed_transaction, indent=2))
    print("------------------- sendTransaction ------------------------")

    headers = {"Content-Type": "application/json"}
    httpx.put(url + "/transactions", headers=headers, json=data, cookies=cookies)

    # websocketを作成する
    def on_message(ws, message):
        json_data = loads(message)
        subscribe_topic = f"confirmedAdded/{address_w}"
        if "uid" in json_data:
            uid = json_data["uid"]
            body = dumps({"uid": uid, "subscribe": subscribe_topic})
            ws.send(body)

        if json_data.get("topic"):
            topic = json_data["topic"]
            if topic == subscribe_topic:
                print("------------------- Success --------------------------------")
                print("Success: Transaction")
                print(dumps(json_data["data"], indent=2))
                print("------------------------------------------------------------")

    ws_url = url.replace("https", "wss")
    ws = websocket.WebSocketApp(
        ws_url + "/ws",
        on_message=on_message,
        cookie=cookie,
    )
    try:
        ws.run_forever()
    except KeyboardInterrupt:
        pass
    except Exception as e:
        print(e)
    finally:
        ws.close()


if __name__ == "__main__":
    aggregate_complete_transaction_mosaic()
